# 05 - Modeling

Train a model on 2016 features to predict the known 2021 turnout, check how
good it actually is on wards it hasn't seen, then use the same approach on
2021 features to forecast 2026.

Input: `data/processed/train_features.csv` (2016 features + known 2021 turnout)
and `data/processed/forecast_features.csv` (2021 features, no target).


In [1]:
import sys
sys.path.insert(0, "..")

import numpy as np
import pandas as pd
import xgboost as xgb
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import GroupShuffleSplit

from src.features import prepare_model_inputs

train_df = pd.read_csv("../data/processed/train_features.csv")
forecast_df = pd.read_csv("../data/processed/forecast_features.csv")
train_df.shape, forecast_df.shape

((4344, 11), (4344, 10))

## Encoding

`prepare_model_inputs` drops `Ward` (just an ID) and `SnapshotYear` (same
value for every row), drops `MunicipalityCode` as a raw feature (too many
unique codes - one-hot or label encoding it lets the model memorise specific
municipalities instead of learning something general), and one-hot encodes
`Province`.


In [2]:
X, ids, groups = prepare_model_inputs(train_df)
y = train_df["Target_Turnout"]

X.columns.tolist()

['RegisteredVoters_prior',
 'Turnout_prior',
 'IsMetro',
 'LogRegisteredVoters_prior',
 'MunicipalityAvgTurnout_prior',
 'SpoiltRatio_prior',
 'Province_Free State',
 'Province_Gauteng',
 'Province_KwaZulu-Natal',
 'Province_Limpopo',
 'Province_Mpumalanga',
 'Province_North West',
 'Province_Northern Cape',
 'Province_Western Cape']

## Train/test split

Using a plain random split here would let wards from the same municipality
end up in both train and test - since `MunicipalityAvgTurnout_prior` is
built from other wards in that municipality, that would leak information
across the split and inflate the score. Splitting by municipality instead
keeps the test set genuinely unseen.


In [3]:
splitter = GroupShuffleSplit(test_size=0.2, n_splits=1, random_state=42)
train_idx, test_idx = next(splitter.split(X, y, groups=groups))

X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]

overlap = set(groups.iloc[train_idx]) & set(groups.iloc[test_idx])
print(f"Train: {len(X_train)} rows, Test: {len(X_test)} rows")
print(f"Municipalities in both train and test: {len(overlap)}")

Train: 3655 rows, Test: 689 rows
Municipalities in both train and test: 0


## Baseline: last election's turnout alone

Worth checking this before the full model - if the full model can't beat just using `Turnout_prior`, the extra features aren't earning their place.

In [4]:
baseline = LinearRegression()
baseline.fit(X_train[["Turnout_prior"]], y_train)
baseline_pred = baseline.predict(X_test[["Turnout_prior"]])

print(f"Baseline R² (Turnout_prior only): {r2_score(y_test, baseline_pred):.3f}")

Baseline R² (Turnout_prior only): 0.373


## Full model

In [5]:
model = xgb.XGBRegressor(objective="reg:squarederror", n_estimators=100, random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

r2 = r2_score(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"R²: {r2:.3f}")
print(f"RMSE: {rmse:.3f} ({rmse*100:.1f} percentage points)")

R²: 0.443
RMSE: 0.064 (6.4 percentage points)


The full model beats the baseline by a modest amount - `Turnout_prior` is
doing most of the work, but the other features (IsMetro, MunicipalityAvgTurnout,
SpoiltRatio, Province) do add some real signal on top. See `06_evaluation.ipynb`
for a closer look at where the model does well and where it doesn't.

## Feature importance

In [6]:
importance = pd.Series(model.feature_importances_, index=X.columns).sort_values(ascending=False)
importance

Turnout_prior                   0.216051
Province_Mpumalanga             0.136248
Province_Limpopo                0.135427
Province_Northern Cape          0.069675
Province_North West             0.067649
RegisteredVoters_prior          0.057873
Province_Gauteng                0.057748
IsMetro                         0.050274
Province_KwaZulu-Natal          0.049965
Province_Western Cape           0.048307
Province_Free State             0.043999
MunicipalityAvgTurnout_prior    0.038400
SpoiltRatio_prior               0.028383
LogRegisteredVoters_prior       0.000000
dtype: float32

## Refit on all data, then forecast 2026

The evaluation above tells us how good the approach is. For the actual
2026 forecast we refit on all of the training data (2016 -> 2021), then
apply that model to the 2021 features to predict 2026 - the real,
unknown outcome.


In [7]:
final_model = xgb.XGBRegressor(objective="reg:squarederror", n_estimators=100, random_state=42)
final_model.fit(X, y)

X_forecast, ids_forecast, _ = prepare_model_inputs(forecast_df, reference_columns=X.columns)
predicted_2026 = final_model.predict(X_forecast)

forecast_results = ids_forecast.copy()
forecast_results["Turnout_2021"] = forecast_df["Turnout_prior"]
forecast_results["Predicted_Turnout_2026"] = predicted_2026
forecast_results.head()

,Province,Ward,MunicipalityCode,Turnout_2021,Predicted_Turnout_2026
0,Eastern Cape,Ward 29200001,BUF,0.400459,0.272586
1,Eastern Cape,Ward 29200002,BUF,0.441672,0.332271
2,Eastern Cape,Ward 29200003,BUF,0.338520,0.238368
3,Eastern Cape,Ward 29200004,BUF,0.491476,0.390433
4,Eastern Cape,Ward 29200005,BUF,0.403802,0.219514


In [8]:
print("Predicted 2026 turnout:")
print(forecast_results["Predicted_Turnout_2026"].describe())
print()
print("Actual 2021 turnout, for comparison:")
print(forecast_results["Turnout_2021"].describe())

Predicted 2026 turnout:
count    4344.000000
mean        0.392913
std         0.091468
min         0.120470
25%         0.346432
50%         0.400203
75%         0.448087
max         0.800639
Name: Predicted_Turnout_2026, dtype: float64

Actual 2021 turnout, for comparison:
count    4344.000000
mean        0.467693
std         0.086488
min         0.119934
25%         0.410007
50%         0.462686
75%         0.522558
max         0.855718
Name: Turnout_2021, dtype: float64


Mean predicted turnout keeps dropping (57.2% in 2016, 46.8% in 2021, now
~39% predicted for 2026) - consistent with the declining trend seen in the
EDA, not a sign of something broken.

## Save the model and forecast

In [9]:
import joblib

joblib.dump({"model": final_model, "columns": list(X.columns)}, "../src/model.pkl")
forecast_results.to_csv("../data/processed/turnout_forecast_2026.csv", index=False)

print("Saved model to src/model.pkl")
print("Saved forecast to data/processed/turnout_forecast_2026.csv")

Saved model to src/model.pkl
Saved forecast to data/processed/turnout_forecast_2026.csv
